### PR TITLE
feat: added account balance migrations

### PR DESCRIPTION
**Pull request by victoraugustolls**
**ATTENTION**

Part of the tests done in this PR should move to [pgTAP](https://pgtap.org/) like testing, and not repository testing. This will be done in a future PR.
- - -

This PR refactors the `get_account_balance` query. Now we have a snapshot each time a account balance is queried. We aggregate over all the account transactions by date, get the second most recent date and save it on a snapshot. This guarantee us that no transaction will be missed when querying the balance.

Although this is done synchronous for now, it should become async as it is not a necessary operation, but a optimization technique, so it should not cause any delay when querying an account balance.

### CODE REVIEW

File: app/gateways/db/postgres/migrations/000003_account_balance.up.sql [hunk 1]
```diff
@@ -0,0 +1,187 @@
+begin;
+
+create table if not exists account_balance
+(
+    credit_balance bigint      not null,
+    debit_balance  bigint      not null,
+    tx_date        timestamptz not null,
+    account        ltree primary key
+) with (fillfactor = 50);
```
Comment (thalissonfelipe) [inline app/gateways/db/postgres/migrations/000003_account_balance.up.sql:9]: Só curiosidade mesmo, o que esse with (fillfactor = 50) faz?

Comment (victoraugustolls) [reply app/gateways/db/postgres/migrations/000003_account_balance.up.sql:9]: Ele diz pro postgres usar só metade de uma página pra inserção de novos dados, a outra metade fica reservada pra update. Como a tendência é ter mais update do que insert ao longo do tempo, isso é uma otimização para updates.

**Review by brunodecastro — APPROVED** [review]

File: app/gateways/db/postgres/migrations/000002_account_versioning.up.sql [hunk 2]
```diff
@@ -0,0 +1,53 @@
+begin;
+
+create table if not exists account_version (
+    version int,
+    account ltree primary key
+) with (fillfactor = 50);
+
+create or replace function update_account_version()
```
Comment (brunodecastro) [inline app/gateways/db/postgres/migrations/000002_account_versioning.up.sql:8]: Show, nós vamos precisar disso no front-api =)

Comment (victoraugustolls) [reply app/gateways/db/postgres/migrations/000002_account_versioning.up.sql:8]: @brunodecastro boaa! so chamar se precisar :D

File: app/domain/vos/account_path.go [hunk 3]
```diff
@@ -56,9 +56,9 @@ var _defaultConfig = AccountConfig{
 //
 // Some examples:
 //   - assets.account.treasury
-//   - liability.available.96a131a8-c4ac-495e-8971-fcecdbdd003a
-//   - liability.available.96a131a8-c4ac-495e-8971-fcecdbdd003a.some_detail
-//   - liability.clients.available.96a131a8-c4ac-495e-8971-fcecdbdd003a.detail1.detail2
+//   - liability.available.96a131a8_c4ac_495e_8971_fcecdbdd003a
+//   - liability.available.96a131a8_c4ac_495e_8971_fcecdbdd003a.some_detail
+//   - liability.clients.available.96a131a8_c4ac_495e_8971_fcecdbdd003a.detail1.detail2
```
Comment (matheusmosca) [inline app/domain/vos/account_path.go:61]: Se eu passar um path com `-` para a função `NewAccountPath`, será retornado um erro?

**Review by matheusmosca — APPROVED** [review]

File: app/domain/vos/account_path.go (hunk 3 above)
Comment (victoraugustolls) [reply app/domain/vos/account_path.go:61]: Not yet, I'm still trying to figure it out where this should happen, as other storage options might be available where `-` character is a valid one. Accepting PR's for this if you are interested.

**Comment by thalissonfelipe** [comment]
> @thalissonfelipe blank identifier you meant "_", right? Yeap, that makes sense! thanks

Yes! Sorry! :P

**Review by thalissonfelipe — APPROVED** [review]

**Review by Streppel — APPROVED** [review]

**Review by matheusmosca — APPROVED** [review]

**Review by alexmourapb — APPROVED** [review]

**Pull request merged**